libass/ass.c

```diff
@@ -261,64 +280,64 @@ static long long string2timecode(ASS_Library *library, char *p)
 #define PARSE_END   }
 
 #define ANYVAL(name,func) \
-    } else if (ass_strcasecmp(tname, #name) == 0) { \
+    } else if (ASS_SV_IEQ(tname, #name)) { \
         target->name = func(token);
 
 #define STRVAL(name) \
-    } else if (ass_strcasecmp(tname, #name) == 0) { \
-        char *new_str = strdup(token); \
+    } else if (ASS_SV_IEQ(tname, #name)) { \
+        char *new_str = ass_copy_string(token); \
         if (new_str) { \
             free(target->name); \
             target->name = new_str; \
         }
 
 #define STARREDSTRVAL(name) \
-    } else if (ass_strcasecmp(tname, #name) == 0) { \
-        while (*token == '*') ++token; \
-        char *new_str = strdup(token); \
+    } else if (ASS_SV_IEQ(tname, #name)) { \
+        while (token.len && *token.str == '*') ++token.str, --token.len; \
+        char *new_str = ass_copy_string(token); \
         if (new_str) { \
             free(target->name); \
             target->name = new_str; \
         }
 
+static float sv_atof(ASS_StringView val)
+{
+    // Header lines always end in a newline or \0, so this can't overflow
```

Comment:
You make a good point that `strtod` can’t work with unterminated `ASS_StringView`. It seems that at this point we don’t _need_ this, but this might come up again in the future.

If and when we do need `ass_strtod` on potentially unterminated `ASS_StringView`, well, I suppose we’ll have no choice but to adapt `ass_strtod`. But I still think we should prefer to stay away from it until it’s really necessary, as it requires nontrivial effort to understand the current `ass_strtod` code to properly patch it and considerably raises the bar for any future effort to switch to another `strtod` implementation (which is likely to have more complex code than the current one).

The number format (for floating-point numbers) is defined as “Windows `wcstod` and/or `swscanf("%f")`”. Which generally means “standard C89 `strtod`”. And our implementation is just an old but run-of-the-mill `strtod` implementation (whose quality on extreme inputs I’m willing to bet differs from VSFilter’s, as it’s much simpler than `strtod` implementations in any serious C library from this century) taken verbatim from some third-party source with `ctype.h` calls replaced by locale-invariant `ass_` versions and, much later, edited to fix a few generic issues independent of any Windows behaviour. I think the only thing we’ve ever compared to VSFilter is numbers with `d` as the exponent separator, to confirm that we don’t need to add support for them as it isn’t accepted by whatever versions of Microsoft’s C libraries VSFilter binaries use. But certainly, if we ever do replace the implementation, we should check that we don’t introduce extra features like hexadecimal numbers (unless by then they become common in VSFilter binaries).

Floating-point parsing is pretty complicated. I don’t think we can afford to maintain a floating-point parser as a tightly integrated, core part of libass like we maintain our integer parsers, which are trivial in comparison. Or perhaps it’s better to say I hope that we don’t have to. Although I suppose we do maintain a rasterizer and blur, which only one person has command of :-) But here, I think none of us would have (lasting) full grasp of this code.

Reply:
> our implementation[’s] quality on extreme inputs I’m willing to bet differs from VSFilter’s

For the record, I’m still willing to bet that our results differ, but Microsoft’s `strtod` [wasn’t always perfect](https://www.exploringbinary.com/visual-c-plus-plus-strtod-still-broken/) either :-) No surprises though. This stuff is hard.